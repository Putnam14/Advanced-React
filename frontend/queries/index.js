import gql from 'graphql-tag';
import { itemDetails } from './fragments';

export const CREATE_ITEM_MUTATION = gql`
  ${itemDetails}
  mutation createItem($description: String!, $title: String!, $price: Int!, $image: String, $largeImage: String) {
    createItem(description: $description, title: $title, price: $price, image: $image, largeImage: $largeImage) {
      ...itemDetails
    }
  }
`;

export const SIGNUP_MUTATION = gql`
  mutation signup($email: String!, $name: String!, $password: String!) {
    signup(email: $email, name: $name, password: $password) {
      token
      user {
        id
        email
        name
      }
    }
  }
`;

export const SIGNIN_MUTATION = gql`
  mutation signup($email: String!, $password: String!) {
    signin(email: $email, password: $password) {
      token
      user {
        id
        email
        name
      }
    }
  }
`;

export const REQUEST_RESET_MUTATION = gql`
  mutation requestReset($email: String!) {
    requestReset(email: $email) {
      id
    }
  }
`;

export const RESET_MUTATION = gql`
  mutation resetPassword($resetToken: String!, $password: String!, $confirmPassword: String!) {
    resetPassword(resetToken: $resetToken, password: $password, confirmPassword: $confirmPassword) {
      token
      user {
        id
        email
        name
      }
    }
  }
`;

export const CREATE_ORDER_MUTATION = gql`
  mutation CreateOrderMutation($token: String!) {
    createOrder(token: $token) {
      id
      charge
      total
      items {
        id
        title
      }
    }
  }
`;

export const ALL_ITEMS_QUERY = gql`
  # Import the Fragment
  ${itemDetails}
  query AllItemsQuery($skip: Int = 0, $first: Int = 2) {
    itemsConnection(orderBy: createdAt_DESC, first: $first, skip: $skip) @connection(key: "itemsConnection") {
      aggregate {
        count
      }
    }
    items(orderBy: createdAt_DESC, first: $first, skip: $skip) @connection(key: "items") {
      ...itemDetails
    }
  }
`;

export const SINGLE_ITEM_QUERY = gql`
  ${itemDetails}
  query FindItem($id: ID!) {
    items(where: { id: $id }) {
      user {
        id
        email
        name
      }
      ...itemDetails
    }
  }
`;

export const SINGLE_ORDER_QUERY = gql`
  query order($id: ID!) {
    order(where: { id: $id }) {
      id
      charge
      total
      createdAt
      items {
        id
        title
        price
        description
        image
        quantity
      }
    }
  }
`;

export const SEARCH_ITEMS_QUERY = gql`
  ${itemDetails}
  query SearchItems($searchTerm: String!) {
    items(
      where: {
        OR: [
          { title_contains: $searchTerm }
          { description_contains: $searchTerm }
        ]
      }
    ) {
      ...itemDetails
    }
  }
`;

export const REMOVE_ITEM_MUTATION = gql`
  mutation RemoveItem($id: ID!) {
    deleteItem(id: $id) {
      id
      title
      description
    }
  }
`;

export const UPDATE_ITEM_MUTATION = gql`
  ${itemDetails}
  mutation UpdateItem(
    $id: ID!
    $title: String
    $description: String
    $price: Int
  ) {
    updateItem(
      id: $id
      description: $description
      title: $title
      price: $price
    ) {
      ...itemDetails
    }
  }
`;

export const UPDATE_USER_MUTATION = gql`
  mutation UpdateUser($name: String!) {
    updateUser(name: $name) {
      name
    }
  }
`;

export const CURRENT_USER_QUERY = gql`
  ${itemDetails}
  query userQuery {
    me {
      id
      email
      name
      orders {
        charge
        id
        total
      }
      cart {
        id
        quantity
        item {
          ...itemDetails
        }
      }
    }
  }
`;

export const USER_ORDERS_QUERY = gql`
  query orders {
    orders(orderBy: createdAt_DESC) {
      id
      total
      createdAt
      updatedAt
      items {
        id
        title
        price
        description
        quantity
        image
      }
    }
  }
`;

export const ADD_TO_CART_MUTATION = gql`
  mutation addToCart($id: ID!) {
    addToCart(id: $id) {
      id
      quantity
      item {
        id
        price
        description
        image
        title
      }
    }
  }
`;

export const REMOVE_FROM_CART_MUTATION = gql`
  mutation removeFromCart($id: ID!) {
    removeFromCart(id: $id) {
      id
    }
  }
`;

export const ALL_USERS_QUERY = gql`
  query Users {
    users {
      id
      name
      email
      permissions
    }
  }
`;

export const UPDATE_PERMISSIONS_MUTATION = gql`
  mutation updatePermissions($permissions: [Permission], $userId: ID!) {
    updatePermissions(permissions: $permissions, userId: $userId) {
      id
      permissions
      name
      email
    }
  }
`;
