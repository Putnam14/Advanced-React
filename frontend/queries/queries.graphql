fragment itemDetails on Item {
  __typename
  id
  title
  price
  description
  image
  largeImage
}

mutation CREATE_ITEM_MUTATION($description: String!, $title: String!, $price: Int!, $image: String, $largeImage: String) {
  createItem(description: $description, title: $title, price: $price, image: $image, largeImage: $largeImage) {
    ...itemDetails
  }
}

mutation SIGNUP_MUTATION($email: String!, $name: String!, $password: String!) {
  signup(email: $email, name: $name, password: $password) {
    token
    user {
      id
      email
      name
    }
  }
}

mutation SIGNIN_MUTATION($email: String!, $password: String!) {
  signin(email: $email, password: $password) {
    token
    user {
      id
      email
      name
    }
  }
}

mutation REQUEST_RESET_MUTATION($email: String!) {
  requestReset(email: $email) {
    id
  }
}

mutation RESET_MUTATION($resetToken: String!, $password: String!, $confirmPassword: String!) {
  resetPassword(resetToken: $resetToken, password: $password, confirmPassword: $confirmPassword) {
    token
    user {
      id
      email
      name
    }
  }
}

mutation CREATE_ORDER_MUTATION($token: String!) {
  createOrder(token: $token) {
    id
    charge
    total
    items {
      id
      title
    }
  }
}

query ALL_ITEMS_QUERY($skip: Int = 0, $first: Int = 4) {
  itemsConnection(orderBy: createdAt_DESC, first: $first, skip: $skip) {
    aggregate {
      count
    }
  }
  items(orderBy: createdAt_DESC, first: $first, skip: $skip) {
    ...itemDetails
  }
}

query SINGLE_ITEM_QUERY($id: ID!) {
  items(where: { id: $id }) {
    user {
      id
      email
      name
    }
    ...itemDetails
  }
}

query SINGLE_ORDER_QUERY($id: ID!) {
  order(id: $id) {
    id
    charge
    total
    createdAt
    user {
      id
    }
    items {
      id
      title
      price
      description
      image
      quantity
    }
  }
}

query SEARCH_ITEMS_QUERY($searchTerm: String!) {
  items(
    where: {
      OR: [
        { title_contains: $searchTerm }
        { description_contains: $searchTerm }
      ]
    }
  ) {
    ...itemDetails
  }
}

mutation REMOVE_ITEM_MUTATION($id: ID!) {
  deleteItem(id: $id) {
    id
    title
    description
  }
}

mutation UPDATE_ITEM_MUTATION(
  $id: ID!
  $title: String
  $description: String
  $price: Int
) {
  updateItem(
    id: $id
    description: $description
    title: $title
    price: $price
  ) {
    ...itemDetails
  }
}

mutation UPDATE_USER_MUTATION($name: String!) {
  updateUser(name: $name) {
    name
  }
}

query CURRENT_USER_QUERY {
  me {
    id
    email
    name
    permissions
    orders {
      id
      charge
      total
    }
    cart {
      id
      quantity
      item {
        ...itemDetails
      }
    }
  }
}

query USER_ORDERS_QUERY {
  orders(orderBy: createdAt_DESC) {
    id
    total
    createdAt
    updatedAt
    items {
      id
      title
      price
      description
      quantity
      image
    }
  }
}

mutation ADD_TO_CART_MUTATION($id: ID!) {
  addToCart(id: $id) {
    id
    quantity
    item {
      id
      price
      description
      image
      title
    }
  }
}

mutation REMOVE_FROM_CART_MUTATION($id: ID!) {
  removeFromCart(id: $id) {
    id
  }
}

query ALL_USERS_QUERY {
  users {
    id
    name
    email
    permissions
  }
}

mutation UPDATE_PERMISSIONS_MUTATION($permissions: [Permission], $userId: ID!) {
  updatePermissions(permissions: $permissions, userId: $userId) {
    id
    permissions
    name
    email
  }
}

query LOCAL_STATE_QUERY {
  cartOpen @client
}

mutation TOGGLE_CART_MUTATION {
  toggleCart @client
}

