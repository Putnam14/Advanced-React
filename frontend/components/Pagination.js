import React from 'react';
import { Query } from 'react-apollo';
import styled from 'styled-components';
import Link from 'next/link';
import PropTypes from 'prop-types';
import { perPage } from '../config';
import { ALL_ITEMS_QUERY } from '../queries/queries.graphql';

const PaginationStyles = styled.div`
  text-align: center;
  display: inline-grid;
  grid-template-columns: repeat(4, auto);
  align-items: stretch;
  justify-content: center;
  align-content: center;
  margin: 2rem 0;
  border: 1px solid ${props => props.theme.lightgrey};
  border-radius: 10px;
  & > * {
    margin: 0;
    padding: 15px 30px;
    border-right: 1px solid ${props => props.theme.lightgrey};
    &:last-child {
      border-right: 0;
    }
  }
  a[aria-disabled='true'] {
    color: grey;
    pointer-events: none;
  }
`;

const Pagination = props => (
  <Query query={ALL_ITEMS_QUERY}>
    {({ data, loading, error }) => {
      if (loading || error) return null;
      const { aggregate } = data.itemsConnection;
      const { page } = props;
      const pages = Math.ceil(aggregate.count / perPage);
      // TODO prefetch this
      return (
        <PaginationStyles data-test="pagination">
          <Link
            prefetch
            href={{
              pathname: 'items',
              query: { page: page - 1 },
            }}
          >
            <a className="prev" aria-disabled={page <= 1}>
              ←Prev
            </a>
          </Link>
          <p>
            Page <strong>{page} </strong> of <strong className="totalPages">{pages}</strong>
          </p>
          <p>
            <strong>{aggregate.count}</strong> Items Total
          </p>
          <Link
            prefetch
            href={{
              pathname: 'items',
              query: { page: page + 1 },
            }}
          >
            <a className="next" aria-disabled={page >= pages}>
              Next →
            </a>
          </Link>
        </PaginationStyles>
      );
    }}
  </Query>
);

Pagination.propTypes = {
  page: PropTypes.number.isRequired,
};

export default Pagination;
