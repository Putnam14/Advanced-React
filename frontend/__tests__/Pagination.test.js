import React from 'react';
import { shallow, mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import Pagination from '../components/Pagination';
import { MockedProvider } from 'react-apollo/test-utils';
import { fakeItem } from './testUtils';
import { ALL_ITEMS_QUERY } from '../queries/queries';

const data = {
  itemsConnection: { aggregate: { count: 18 } },
  items: [fakeItem()],
};

describe('<Pagination/>', () => {
  fit('displays loading message', () => {
    const mocks = [
      {
        request: { query: ALL_ITEMS_QUERY, variables: { skip: 0 } },
        delay: 50,
        result: { data },
      },
    ];

    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <Pagination page={1} />
      </MockedProvider>
    );
    console.log(wrapper.debug());
    // expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it('renders pagination for 18 items', () => {
    const wrapper = shallow(<Pagination loading={false} page={1} itemsQuery={fakeQuery} />);
    expect(toJSON(wrapper)).toMatchSnapshot();
    expect(wrapper.find('.totalPages').text()).toEqual('2');
  });

  it('renders pagination for 28 items', () => {
    const fakeQuery2 = {
      itemsConnection: { aggregate: { count: 28 } },
    };
    const wrapper = shallow(<Pagination loading={false} page={1} itemsQuery={fakeQuery2} />);
    expect(wrapper.find('.totalPages').text()).toEqual('4');
  });

  it('disables and enables next/prev buttons', () => {
    const fakeQuery3 = {
      itemsConnection: { aggregate: { count: 100 } },
    };
    const wrapper = shallow(<Pagination loading={false} page={1} itemsQuery={fakeQuery} />);

    expect(wrapper.find('a.prev').props()['aria-disabled']).toEqual(true);
    expect(wrapper.find('a.next').props()['aria-disabled']).toEqual(false);
    wrapper.setProps({ page: 2 });
    expect(wrapper.find('a.prev').props()['aria-disabled']).toEqual(false);
    expect(wrapper.find('a.next').props()['aria-disabled']).toEqual(true);
    // when in the middle, both should work
    wrapper.setProps({ itemsQuery: fakeQuery3, page: 3 });
    expect(wrapper.find('a.prev').props()['aria-disabled']).toEqual(false);
    expect(wrapper.find('a.next').props()['aria-disabled']).toEqual(false);
  });
});
