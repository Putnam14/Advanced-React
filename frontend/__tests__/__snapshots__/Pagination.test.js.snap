// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination/> renders pagination for 18 items 1`] = `
<styled.div>
  <Link
    href={
      Object {
        "pathname": "items",
        "query": Object {
          "page": 0,
        },
      }
    }
    prefetch={true}
  >
    <a
      aria-disabled={true}
      className="prev"
    >
      ←Prev
    </a>
  </Link>
  <p>
    Page 
    <strong>
      1
       
    </strong>
     of 
    <strong
      className="totalPages"
    >
      2
    </strong>
  </p>
  <p>
    <strong>
      18
    </strong>
     Items Total
  </p>
  <Link
    href={
      Object {
        "pathname": "items",
        "query": Object {
          "page": 2,
        },
      }
    }
    prefetch={true}
  >
    <a
      aria-disabled={false}
      className="next"
    >
      Next →
    </a>
  </Link>
</styled.div>
`;
